merkle-tree/merkle_tree.go

```diff
@@ -0,0 +1,226 @@
+		// it's a leaf node
+		dataToHash = data
+	} else {
+		// it's an inner node
+		dataToHash = append(left.hash.Bytes(), right.hash.Bytes()...)
+	}
+
+	hasher.Reset()
+	hasher.Write(dataToHash)
+	node.hash = types.BytesToHash(hasher.Sum(nil))
+
+	return node
+}
+
+// MerkleTree is the structure for the Merkle tree.
+type MerkleTree struct {
+	// hasher is a pointer to the hashing struct (e.g., Keccak256)
+	hasher hash.Hash
+	// rootNode is the root node of the tree
+	rootNode *MerkleNode
+	// leaf nodes is the list of leaf nodes of the tree (the lowest level of the tree)
+	leafNodes []*MerkleNode
+}
+
+// NewMerkleTree creates a new Merkle tree from the provided data and using the default hashing (Keccak256).
+func NewMerkleTree(data [][]byte) (*MerkleTree, error) {
+	return NewMerkleTreeWithHashing(data, crypto.NewKeccakState())
+}
+
+// NewMerkleTreeWithHashing creates a new Merkle tree from the provided data and hash type
+func NewMerkleTreeWithHashing(data [][]byte, hasher hash.Hash) (*MerkleTree, error) {
+	if len(data) == 0 {
+		return nil, errors.New("tree must contain at least one leaf")
+	}
+
+	leafNodes := make([]*MerkleNode, len(data))
+	for i, d := range data {
+		leafNodes[i] = newMerkleNode(nil, nil, d, hasher)
+	}
+
+	nodes := leafNodes
+	for len(nodes) > 1 {
+		var newLevel []*MerkleNode
+
+		for i := 0; i < len(nodes); i += 2 {
+			left := nodes[i]
+			right := nodes[i] // if the tree has uneven number of nodes, this will duplicate the last one
+
+			if i+1 < len(nodes) {
+				right = nodes[i+1] // if it has even numbers, then this will point to the last node
+			}
+
+			parent := newMerkleNode(left, right, nil, hasher)
+			left.parent = parent
+			right.parent = parent
+
+			newLevel = append(newLevel, parent)
+		}
+
+		nodes = newLevel
+	}
+
+	return &MerkleTree{
+		hasher:    hasher,
+		rootNode:  nodes[0],
+		leafNodes: leafNodes,
+	}, nil
+}
+
+// LeafIndex returns the index of given leaf if found in tree
+func (t *MerkleTree) LeafIndex(leaf []byte) (uint64, error) {
+	for i, leafNode := range t.leafNodes {
+		if bytes.Equal(leafNode.data, leaf) {
+			return uint64(i), nil
+		}
+	}
+
+	return 0, errLeafNotFound
+}
+
+// Hash is the Merkle Tree root hash
+func (t *MerkleTree) Hash() types.Hash {
+	return t.rootNode.hash
+}
+
+// String implements the stringer interface
+func (t *MerkleTree) String() string {
+	return hex.EncodeToString(t.Hash().Bytes())
+}
+
+// Depth returns the depth of merkle tree
+func (t *MerkleTree) Depth() int {
+	return int(math.Ceil(math.Log2(float64(len(t.leafNodes)))))
+}
+
+// GenerateProof generates the proof of membership for a piece of data in the Merkle tree.
+func (t *MerkleTree) GenerateProof(leaf []byte) ([]types.Hash, error) {
+	proof := []types.Hash{}
+
+	leafNode := t.findLeafNode(leaf)
+	if leafNode == nil {
+		return nil, fmt.Errorf("given data not in merkle tree")
+	}
+
+	node := leafNode
+	for !bytes.Equal(node.hash.Bytes(), t.rootNode.hash.Bytes()) {
```

Comment:
If I'm correct, you could compare it this way as well (but double check before applying suggestion) :slightly_smiling_face: 
```suggestion
	for node.hash != t.rootNode.hash {
```